### PR TITLE
Recover accuracy

### PR DESCRIPTION
**Pull request by chris-ha458**
If `mess_difference <0.01`
we see if `coherence_difference > 0.02` and return partialord based on that.
If not, we try to use multibyte usage difference if it is big enough.

comparing with ` multibyte_usage_a.abs() > f32::epsilon` is idiomatic and includes when the value is 0.0 or some value very close to it.

However, it does not change the final accuracy at all.

### CODE REVIEW

**Comment by chris-ha458** [comment]
each commit represents different ways to represent the same idea, but none make a difference.

**Comment by nickspring** [comment]
If think it is not correct. The situation when multibyte_a = 0 and multibyte_b != 0 is totally correct. We just shouldn't do decision if multibyte_a == multibyte_b for example (0 and 0, 3 and 3, etc) (mess should be used).

File: src/entity.rs [hunk 1]
```diff
@@ -124,8 +124,16 @@ impl PartialOrd<Self> for CharsetMatch {
...
-        if mess_difference < 0.01 && coherence_difference > 0.02 {
-            return coherence_b.partial_cmp(&coherence_a);
+        if mess_difference < 0.01 {
+            if coherence_difference > 0.02 {
+                return coherence_b.partial_cmp(&coherence_a);
+            }
+            let multibyte_usage_a = self.multi_byte_usage();
+            let multibyte_usage_b = other.multi_byte_usage();
+            let multibyte_usage_delta = (multibyte_usage_a - multibyte_usage_b).abs();
+            if multibyte_usage_a.abs() > f32::EPSILON && multibyte_usage_b.abs() > f32::EPSILON {
```
Comment (nickspring) [inline src/entity.rs:134]: maybe it should be `multibyte_usage_delta > f32:EPSILON ` ?

File: src/entity.rs [hunk 2]
```diff
@@ -124,8 +124,16 @@ impl PartialOrd<Self> for CharsetMatch {
...
-        if mess_difference < 0.01 && coherence_difference > 0.02 {
-            return coherence_b.partial_cmp(&coherence_a);
+        if mess_difference < 0.01 {
+            if coherence_difference > 0.02 {
+                return coherence_b.partial_cmp(&coherence_a);
+            }
+            let multibyte_usage_a = self.multi_byte_usage();
+            let multibyte_usage_b = other.multi_byte_usage();
+            let multibyte_usage_delta = (multibyte_usage_a - multibyte_usage_b).abs();
+            if multibyte_usage_delta > f32::EPSILON {
```
Comment (chris-ha458) [inline src/entity.rs]: different comparison

File: src/entity.rs (hunk 1 above)
Comment (chris-ha458) [reply src/entity.rs:134]: https://github.com/nickspring/charset-normalizer-rs/pull/25/commits/c3136ec4f44573adc693ece90d1deb213febc909#r1342504166

unfortunately that does not work either.

**Comment by chris-ha458** [comment]
> If think it is not correct. The situation when multibyte_a = 0 and multibyte_b != 0 is totally correct. We just shouldn't do decision if multibyte_a == multibyte_b for example (0 and 0, 3 and 3, etc) (mess should be used).
i'm not sure if i fully understand what you mean. Can you show me code or maybe pseudocode(if else) what you mean?

File: src/entity.rs (hunk 1 above)
Comment (nickspring) [reply src/entity.rs:134]: What do you mean "does not work"?
Accuracy: 97.1% that's OK.

File: src/entity.rs (hunk 2 above)
Comment (nickspring) [reply src/entity.rs]: what is your expectation for this? :) 97.1% is our goal, your code has passed all checks.

File: src/entity.rs (hunk 1 above)
Comment (chris-ha458) [reply src/entity.rs:134]: Ah I meant that it does not reach 97.1%

Comment (nickspring) [reply src/entity.rs:134]: Both versions have 97.1%, but I wrote just about logic - we should compare multibyte_usage values if they are different only.

Comment (chris-ha458) [reply src/entity.rs]: Oh. Now I understand.
There must be an issue on my end! That is unfortunate.
Reverted to  `multibyte_usage_delta > f32:EPSILON`
in d4eae53

**Comment by chris-ha458** [comment]
If i Understand correctly, on your system the final accuracy results in 97.1%?
using  
`cargo run --release --bin performance --all-features |tail -n 50`
my system shows a result of 96.8%

```
--> A) charset-normalizer-rs Conclusions
   --> Accuracy: 96.8%
   --> Total time: 642.285389ms
   --> Avg time: 1.570379ms
   --> 50th: 662.846µs
   --> 95th: 4.530332ms
   --> 99th: 11.080102ms
```

I'm I checking this right? Is your system showing higher than 97.0% under same code?

**Comment by nickspring** [comment]
You could always check accuracy and speed in the output of performance action
https://github.com/nickspring/charset-normalizer-rs/actions/runs/6378885659/job/17310431300?pr=25
I see 97.1% here and I have 97.1% locally. What OS do you have?

**Comment by chris-ha458** [comment]
Ah you are correct.

I am using WSL2 but I plan to setup Windows and Linux (via virtualbox) workflows.

**Comment by nickspring** [comment]
Interesting :) maybe for this platform `encoding` library offers fewer encodings...

**Pull request merged**